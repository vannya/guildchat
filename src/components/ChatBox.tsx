import React, { useState, SyntheticEvent, ChangeEvent } from 'react';
import { MessageForm } from './MessageForm';
import { MessageList } from './MessageList';
import { ChatBoxHeader } from './ChatBoxHeader';
import { MessageItem } from '../types';

interface IChatBoxProps {
	receiverName: string;
	messages: MessageItem[];
	onSubmit: (text: string) => void;
}

export const ChatBox = ({
	receiverName,
	messages,
	onSubmit
}: IChatBoxProps): JSX.Element => {
	const [text, setText] = useState<string>('');

	const onClickHandler = (e: SyntheticEvent): void => {
		e.preventDefault();
		const trimmedText = text.trim();
		if (trimmedText !== '') {
			onSubmit(trimmedText);
		}
		setText('');
	};

	const onChangeHandler = (e: ChangeEvent<HTMLInputElement>): void => {
		setText(e.target.value);
	};
	return (
		<div>
			<ChatBoxHeader userName={receiverName} />
			<div className='chat-box'>
				{messages && <MessageList messages={messages} />}
				<MessageForm
					id='chat-message'
					value={text}
					disabled={!text.trim()}
					onChange={onChangeHandler}
					onClick={onClickHandler}
				/>
			</div>
		</div>
	);
};
