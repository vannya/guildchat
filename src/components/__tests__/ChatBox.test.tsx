import React from 'react';
import { cleanup, render } from '@testing-library/react';
import userEvents from '@testing-library/user-event';
import { ChatBox, IChatBoxProps } from '../ChatBox';

// Mock scrollIntoView for MessageList child
const scrollIntoViewMock = jest.fn();
window.HTMLElement.prototype.scrollIntoView = scrollIntoViewMock;

describe('ChatBox', () => {
	const getProps = (): IChatBoxProps => {
		return {
			receiverName: 'Test Name',
			messages: [],
			onSubmit: jest.fn()
		};
	};

	it('should match snapshot', () => {
		const props = getProps();
		const { container } = render(<ChatBox {...props} />);
		expect(container).toMatchSnapshot();
		cleanup();
	});

	it('should update the text', () => {
		const props = getProps();
		const { container } = render(<ChatBox {...props} />);
		// Change text
		const inputField = container.querySelectorAll('#chat-message');
		expect(inputField).toHaveLength(1);
		expect(inputField[0].getAttribute('value')).toBe('');
		// Update text
		userEvents.type(inputField[0], 'Message to send');
		expect(inputField[0].getAttribute('value')).toBe('Message to send');
		cleanup();
	});

	it('should submit text', () => {
		const props = getProps();
		const { container, getAllByText } = render(<ChatBox {...props} />);
		// Change text
		const inputField = container.querySelectorAll('#chat-message');
		expect(inputField).toHaveLength(1);
		userEvents.type(inputField[0], 'Message to send');
		// Submit
		const submitButton = getAllByText('Send');
		expect(submitButton).toHaveLength(1);
		userEvents.click(submitButton[0]);

		expect(props.onSubmit).toHaveBeenCalledTimes(1);
		expect(props.onSubmit).toHaveBeenCalledWith('Message to send');
		cleanup();
	});

	it('should submit trimmed text', () => {
		const props = getProps();
		const { container, getAllByText } = render(<ChatBox {...props} />);
		// Change text
		const inputField = container.querySelectorAll('#chat-message');
		expect(inputField).toHaveLength(1);
		userEvents.type(inputField[0], '    Message to send   ');
		// Submit
		const submitButton = getAllByText('Send');
		expect(submitButton).toHaveLength(1);
		expect(submitButton[0].getAttribute('disabled')).toBe(null);
		userEvents.click(submitButton[0]);

		expect(props.onSubmit).toHaveBeenCalledTimes(1);
		expect(props.onSubmit).toHaveBeenCalledWith('Message to send');
		cleanup();
	});

	it('should not submit empty text', () => {
		const props = getProps();
		const { container, getAllByText } = render(<ChatBox {...props} />);
		// Change text
		const inputField = container.querySelectorAll('#chat-message');
		expect(inputField).toHaveLength(1);
		userEvents.type(inputField[0], '            ');
		// Submit
		const submitButton = getAllByText('Send');
		expect(submitButton).toHaveLength(1);
		userEvents.click(submitButton[0]);

		expect(props.onSubmit).not.toHaveBeenCalledTimes(1);
		cleanup();
	});

	it('should disable the submit button when there is no text', () => {
		const props = getProps();
		const { getAllByText } = render(<ChatBox {...props} />);
		// Submit without adding input value
		const submitButton = getAllByText('Send');
		expect(submitButton).toHaveLength(1);
		userEvents.click(submitButton[0]);
		expect(submitButton[0].getAttribute('disabled')).toBe('');

		expect(props.onSubmit).not.toHaveBeenCalled();
		cleanup();
	});
});
