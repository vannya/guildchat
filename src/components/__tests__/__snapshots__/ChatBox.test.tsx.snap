// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChatBox should match snapshot 1`] = `
<div>
  <div>
    <h2
      class="chatbox-header"
    >
      Chatting with 
      Test Name
    </h2>
    <div
      class="chat-box"
    >
      <div
        class="message-list"
      >
        <div />
      </div>
      <form
        class="message-form"
      >
        <input
          id="chat-message"
          maxlength="500"
          placeholder="Enter Text"
          type="text"
          value=""
        />
        <button
          class="message-form--button"
          disabled=""
        >
          Send
        </button>
      </form>
    </div>
  </div>
</div>
`;
